In [1]:
import seaborn as sns

In [2]:
df = sns.load_dataset('iris')

In [3]:
df.head(3)

,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,setosa
1,4.9,3.0,1.4,0.2,setosa
2,4.7,3.2,1.3,0.2,setosa


In [4]:
df.species.unique()

array(['setosa', 'versicolor', 'virginica'], dtype=object)

In [8]:
df.shape

(150, 5)

In [9]:
from sklearn.preprocessing import LabelEncoder

In [10]:
le =LabelEncoder()

In [11]:
df['species'] = le.fit_transform(df['species'])

In [12]:
df.head(2)

,sepal_length,sepal_width,petal_length,petal_width,species
0,5.1,3.5,1.4,0.2,0
1,4.9,3.0,1.4,0.2,0


In [13]:
df.species.unique()

array([0, 1, 2])

In [15]:
#data split inyto X,Y
X = df.drop(columns=['species'])
y = df['species']

In [16]:
#Data split into train and test

In [17]:
from sklearn.model_selection import train_test_split

In [18]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.20,random_state=11)

Socho 100 students hain aur aap randomly 80 ko training aur 20 ko testing ke liye select kar rahe ho.

Without random_state → har baar students change ho sakte hain.
random_state=11 → ek particular random selection fix ho jayega.
random_state=20 → doosra selection fix ho jayega.

Important: 11 ka koi special meaning nahi hai. Aap 0, 1, 42, 100 etc. bhi use kar sakte ho.

👉 Short definition:
Random state = random splitting ko reproducible/fixed banane ka seed.

In [19]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC 

In [20]:
models = { 
    "Logistic Regression" : LogisticRegression(),
    "Decision Tree Classifier" : DecisionTreeClassifier(),
    "Random Forest Classifier" : RandomForestClassifier(),
    "K-Neighbors Classifier" : KNeighborsClassifier(),
    "Support Vector Classifier" :SVC()
    }

In [21]:
from sklearn.metrics import accuracy_score, confusion_matrix

In [22]:
from sklearn.metrics import accuracy_score,confusion_matrix

In [24]:
for model_name,model in models.items():
    model.fit(X_train,y_train)
    y_pred = model.predict(X_test)
    print(model_name,'\nAccuracy Score :',round(accuracy_score(y_test,y_pred)*100,2))
    print("\nConfusion Matrix\n",confusion_matrix(y_test,y_pred))
    print("---------------------------------------------------")

Logistic Regression 
Accuracy Score : 93.33

Confusion Matrix
 [[ 9  0  0]
 [ 0 10  0]
 [ 0  2  9]]
---------------------------------------------------
Decision Tree Classifier 
Accuracy Score : 93.33

Confusion Matrix
 [[ 9  0  0]
 [ 0 10  0]
 [ 0  2  9]]
---------------------------------------------------
Random Forest Classifier 
Accuracy Score : 93.33

Confusion Matrix
 [[ 9  0  0]
 [ 0 10  0]
 [ 0  2  9]]
---------------------------------------------------
K-Neighbors Classifier 
Accuracy Score : 96.67

Confusion Matrix
 [[ 9  0  0]
 [ 0 10  0]
 [ 0  1 10]]
---------------------------------------------------
Support Vector Classifier 
Accuracy Score : 93.33

Confusion Matrix
 [[ 9  0  0]
 [ 0 10  0]
 [ 0  2  9]]
---------------------------------------------------


In [25]:
#KNN -> Lazy_Model

1. Seaborn Import
import seaborn as sns

Meaning: Seaborn library ko import kar rahe hain.

seaborn → data visualization/datasets ke liye library
sns → seaborn ka short name

👉 Simple: Seaborn ko Python mein le aaye.

2. Iris Dataset Load
df = sns.load_dataset('iris')

Meaning: Seaborn ka ready-made Iris dataset load kar rahe hain.

df → DataFrame jisme dataset store hoga.

Iris dataset mein flowers ki information hai:

sepal_length
sepal_width
petal_length
petal_width
species

👉 Simple: Flower ka data df mein store kar diya.

3. First 3 Rows Dekhna
df.head(3)

Meaning: Dataset ki starting ki 3 rows show karega.

👉 head() = starting rows dekhna.

Example:

5.1  3.5  1.4  0.2  setosa
4.9  3.0  1.4  0.2  setosa
4.7  3.2  1.3  0.2  setosa

👉 Simple: Dataset ke starting ke 3 records check kiye.

4. Unique Species Check
df.species.unique()

Output:

['setosa' 'versicolor' 'virginica']

Meaning: species column mein kaunsi-kaunsi different values hain, ye batayega.

👉 Simple: Flower ki 3 categories hain:

Setosa
Versicolor
Virginica
5. Dataset ka Shape
df.shape

Output:

(150, 5)

Iska meaning:

150 = Rows
5   = Columns

👉 Simple: Dataset mein 150 flowers aur 5 columns hain.

6. LabelEncoder Import
from sklearn.preprocessing import LabelEncoder

Meaning: LabelEncoder ko sklearn se import kiya.

LabelEncoder kya karta hai?

Text categories ko numbers mein convert karta hai.

Example:

setosa      → 0
versicolor  → 1
virginica   → 2

👉 Simple: Machine Learning model ko text samajhne mein problem ho sakti hai, isliye text ko numbers mein convert karte hain.

7. LabelEncoder ka Object
le = LabelEncoder()

Meaning: LabelEncoder ka ek object banaya.

le → LabelEncoder ko represent karega.

👉 Simple: Encoding karne ke liye encoder ready kar diya.

8. Species ko Numbers mein Convert
df['species'] = le.fit_transform(df['species'])

Yahan important hai:

fit_transform()

Ye 2 kaam karta hai:

1. fit → categories ko identify karta hai.

setosa
versicolor
virginica

2. transform → unhe numbers mein convert karta hai.

setosa → 0
versicolor → 1
virginica → 2

👉 Simple: Species ke text ko numerical values mein convert kar diya.

9. Data Dobara Check
df.head(2)

Starting ki 2 rows check ki.

Ab:

sepal_length  sepal_width  petal_length  petal_width  species
5.1           3.5          1.4           0.2          0
4.9           3.0          1.4           0.2          0

Pehle:

setosa

tha.

Ab:

0

hai.

👉 Simple: Encoding successfully hui ya nahi, check kiya.

10. Unique Values Dobara Check
df.species.unique()

Output:

[0 1 2]

👉 Matlab species ab numbers mein convert ho chuki hai.

11. X aur y Banana
X = df.drop(columns=['species'])
y = df['species']

Ye Machine Learning ka bahut important step hai.

X
X = df.drop(columns=['species'])

species ko remove kar diya.

X mein ab:

sepal_length
sepal_width
petal_length
petal_width

hain.

👉 X = Input/Features

y
y = df['species']

species ko target/output bana diya.

👉 y = Output/Target

Real-life example:

Aapke paas:

Height
Weight
Age

se kisi person ka:

Gender

predict karna hai.

Then:

X = Height, Weight, Age
y = Gender

Isi tarah yahan:

X = Flower measurements
y = Flower species

12. train_test_split Import
from sklearn.model_selection import train_test_split

Meaning: Dataset ko training aur testing parts mein divide karne ka function import kiya.

👉 Simple: Data ko Train + Test mein divide karne ke liye function laye.

13. Train-Test Split
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.20,
    random_state=11
)

Ye line bahut important hai.

X

Input data.

y

Output data.

test_size=0.20

20% data testing ke liye.

Remaining:

80% → Training
20% → Testing

150 rows hain:

120 → Training
30  → Testing
random_state=11

Random splitting ko fixed/reproducible banata hai.

Agar same code dobara run karoge, same split milega.

👉 Simple: Data ko 80% training aur 20% testing mein divide kiya aur random selection ko fixed kar diya.

14. 5 ML Models Import
from sklearn.linear_model import LogisticRegression

Logistic Regression model import.

👉 Classification ke liye use hota hai.

from sklearn.tree import DecisionTreeClassifier

Decision Tree model import.

👉 Decision/tree rules ke basis par classification karta hai.

from sklearn.ensemble import RandomForestClassifier

Random Forest model import.

👉 Multiple decision trees ka combination.

from sklearn.neighbors import KNeighborsClassifier

KNN model import.

👉 Nearby/similar data points dekhkar prediction karta hai.

from sklearn.svm import SVC

SVM model import.

👉 Classes ke beech best boundary find karne ki koshish karta hai.

15. Models ko Dictionary mein Store Karna
models = {
    "Logistic Regression": LogisticRegression(),
    "Decision Tree Classifier": DecisionTreeClassifier(),
    "Random Forest Classifier": RandomForestClassifier(),
    "K-Neighbors Classifier": KNeighborsClassifier(),
    "Support Vector Classifier": SVC()
}

Yahan 5 models ko ek dictionary models mein store kiya.

Why?

Taaki same code se saare models ko ek-ek karke train aur test kar sakein.

👉 Simple: 5 ML models ko ek box/dictionary mein rakh diya.

16. Evaluation Metrics Import
from sklearn.metrics import accuracy_score, confusion_matrix

Do metrics import kiye:

accuracy_score

Batata hai model ki prediction kitni correct hai.

Example:

100 predictions
95 correct

Accuracy = 95%
confusion_matrix

Batata hai model ne kaunsi class ko correctly/incorrectly predict kiya.

👉 Simple:

accuracy_score → Kitna correct?
confusion_matrix → Kahan correct/wrong?
17. Main Loop
for model_name, model in models.items():

Dictionary ke har model ko one-by-one nikala ja raha hai.

Example:

1. Logistic Regression
2. Decision Tree
3. Random Forest
4. KNN
5. SVM

👉 Simple: 5 models ko ek-ek karke process karo.

18. Model Training
model.fit(X_train, y_train)

fit() = model ko training data se learn/train karna.

Model:

X_train → Inputs
y_train → Correct Answers

se relationship seekhta hai.

👉 Simple: Model ko padha rahe hain.

Real-life:

Teacher:

Question → Answer

dikha kar student ko train karta hai.

Same way:

X_train → y_train

model ko training dete hain.

19. Prediction
y_pred = model.predict(X_test)

Ab trained model ko unseen/test data diya.

X_test → Model → y_pred

y_pred = model ki predicted values.

👉 Simple: Model se exam/test karwa rahe hain.

20. Accuracy Calculate
accuracy_score(y_test, y_pred)

Compare:

y_test = Actual answer
y_pred = Model ka answer

Example:

Actual:    0 1 2
Predicted: 0 1 2

Jitni predictions correct hongi, accuracy utni high hogi.

21. Accuracy × 100
accuracy_score(y_test, y_pred) * 100

Accuracy normally decimal mein aa sakti hai:

0.9667

×100 karne par:

96.67%
22. Round
round(accuracy_score(y_test, y_pred) * 100, 2)

round(..., 2) → answer ko 2 decimal places tak rakhta hai.

Example:

96.666666

becomes:

96.67
23. Accuracy Print
print(model_name,
      '\nAccuracy Score :',
      round(accuracy_score(y_test,y_pred)*100,2))

Model ka naam aur accuracy print karta hai.

Output example:

K-Neighbors Classifier
Accuracy Score : 96.67
24. Confusion Matrix
confusion_matrix(y_test, y_pred)

Actual aur predicted values ko compare karta hai.

Output for KNN:

[[ 9  0  0]
 [ 0 10  0]
 [ 0  1 10]]

Iska simple meaning:

Class 0 → 9 correct
Class 1 → 10 correct, 1 incorrectly predicted
Class 2 → 10 correct

👉 Simple: Confusion Matrix model ki detailed performance batati hai.

25. Separator Print
print("---------------------------------------------------")

Sirf output ko clean/readable banane ke liye line print kar raha hai.

👉 Simple: Models ke results ko separate karne ke liye.

26. Final Results

Notebook mein results aaye:

Model	Accuracy
Logistic Regression	93.33%
Decision Tree	93.33%
Random Forest	93.33%
KNN	96.67%
SVM	93.33%

Best model:

KNN = 96.67%

Is particular train/test split par KNN ne highest accuracy di.

27. KNN → Lazy Model

Last mein comment hai:

#KNN -> Lazy_Model

KNN ko commonly Lazy Learning Model kaha jata hai.

Why?

KNN training ke time par generally model ko complex rules learn nahi karta.

Instead, prediction ke time:

New data
   ↓
Nearest data points find
   ↓
Unke basis par class decide

👉 Simple: KNN actual decision prediction ke time karta hai, isliye ise Lazy Model kaha jata hai.

🔥 Puri File ka Flow Ek Line Mein

Aapki poori notebook basically ye kaam kar rahi hai:

Iris Dataset
      ↓
Dataset Load
      ↓
Data Check
      ↓
Species ko Number mein Convert
      ↓
X aur y Create
      ↓
80% Train + 20% Test
      ↓
5 Classification Models
      ↓
Models Train
      ↓
Prediction
      ↓
Accuracy Check
      ↓
Confusion Matrix
      ↓
Best Model = KNN (96.67%)

🧠 Beginner ke liye sabse important 7 concepts
1. Dataset       → Data
2. X             → Input/Features
3. y             → Output/Target
4. Train Data    → Model ko sikhana
5. Test Data     → Model ko check karna
6. Accuracy      → Kitna correct
7. Confusion Matrix → Kis class mein kya prediction hua

Overall: Ye notebook ek Classification problem solve kar rahi hai, jahan flower ke measurements (X) se flower ki species (y) predict ki ja rahi hai.